<a href="https://colab.research.google.com/github/christianzambra/productmanagerslife/blob/main/AI_AGENT_BOVESPA_EXP2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Library - Bibliotecas

#Google GenerativeAI library
# if not installed, uncomment the install command below
# !pip install -q -U google-generativeai
import google.generativeai as genai

In [ ]:
#Library - Bibliotecas
#Yahoo Finance - Yahoo Finanças, ainda funciona.
import yfinance as yf
#Pandas - Para poder trabalhar com bases de dados
import pandas as pd

In [ ]:
#Library - Google Workplace - Spreadsheets / Planilhas

from google.colab import auth
import gspread
from google.auth import default

In [ ]:
# Gerando a chave
# 1. Go to Gemini AI studio at https://ai.google.dev/aistudio
# 2. Sign In to Google AI Studio (or create a google account to signin)
# 3. Click on "Get API key"
# 4. Click in "create an API key"
# 5. Copy and save in a safe place
# 6. Create a secure code to put the key in a variable called GOOGLE_API_KEY)

In [ ]:
# TO MAKE THE CODE OPERATES, PLEASE CREATE A FUNCTION TO TAKE YOUR KEY AND LOAD TO THE VARIABLE GOOGLE_API_KEY

In [ ]:
# Start the generative AI with the Google API
#GOOGLE_API_KEY = "PASTE_YOUR_KEY_HERE"
genai.configure(api_key=GOOGLE_API_KEY)

In [ ]:
#Define the model to be used
model = genai.GenerativeModel('gemini-pro')

In [ ]:


# Get Ibovespa data
ticker = yf.Ticker("^BVSP")
data = ticker.history(period="2y")  # Adjust the period as needed

# Print the data
#print(data)

In [ ]:
#print(data)
#print(data[(data.index >= pd.to_datetime("2023-08-01", utc=True).tz_convert("America/Sao_Paulo")) & (data.index <= pd.to_datetime("2023-08-31", utc=True).tz_convert("America/Sao_Paulo"))])

In [ ]:
#Define a Prompt

PRE_PROMPT = f"""
Here is the relevant data for your analysis:

{data.to_markdown(index=False, numalign="left", stralign="left")}

Based on this data, answer the question about IBOVESPA
"""

HUMAN_PROMPT = "What was the average close value for Ibovespa in August 2023?"

POST_PROMPT = "Idenfify and calculate IBOVESPA according to the data, and search on the web for explanation for the variation. Return the answer in 2 paragraphs: The first with the direct response to the question, citing the analysis period, and the second with the explanation for the variaton"


PROMPT = PRE_PROMPT + HUMAN_PROMPT + POST_PROMPT


In [ ]:
#Call the model. The timeout can be configured in seconds, below
response = model.generate_content(PROMPT,
                                  request_options={"timeout": 300})

#This variable has only the mondel answer
resposta = response.text

In [ ]:
print(resposta)

The average closing value for IBOVESPA in August 2023 was 129,266.5. This is based on data from August 1st to 31st, 2023.

The IBOVESPA is a stock market index that tracks the performance of the 100 most traded stocks on the B3 stock exchange in São Paulo, Brazil. It is a widely used benchmark for the Brazilian stock market and is considered a barometer of the Brazilian economy.

The IBOVESPA rose by 4.7% in August 2023, outperforming the S&P 500 index, which rose by 3.1% over the same period. One possible explanation for this outperformance is the relative strength of the Brazilian economy. Brazil's GDP is expected to grow by 1.5% in 2023, which is higher than the IMF's forecast of 2.7% growth for the global economy. Additionally, the Brazilian central bank has been raising interest rates in an effort to curb inflation, which has helped to stabilize the Brazilian real and make it more attractive to foreign investors.


In [ ]:
# This was the first answer of the code, just to be recorded:

#The average closing value for IBOVESPA in August 2023 was 129,266.5. This is based on data from August 1st to 31st, 2023.

#The IBOVESPA is a stock market index that tracks the performance of the 100 most traded stocks on the B3 stock exchange in São Paulo, Brazil. It is a widely used benchmark for the Brazilian stock market and is considered a barometer of the Brazilian economy.

#The IBOVESPA rose by 4.7% in August 2023, outperforming the S&P 500 index, which rose by 3.1% over the same period. One possible explanation for this outperformance is the relative strength of the Brazilian economy. Brazil's GDP is expected to grow by 1.5% in 2023, which is higher than the IMF's forecast of 2.7% growth for the global economy. Additionally, the Brazilian central bank has been raising interest rates in an effort to curb inflation, which has helped to stabilize the Brazilian real and make it more attractive to foreign investors.

Não consigo resistir, vou fazer uma função simples pra checar os dados: 😊

In [ ]:
# I can´t resist and will check the data :)

In [ ]:
# Access S&P 500 index data
sp_ticker = yf.Ticker("^GSPC")  # Use "^GSPC" for S&P 500

# Get historical data for the past 2 years
sp_data = sp_ticker.history(period="2y")

In [ ]:
sp_filtered_data_july = sp_data[(sp_data.index >= pd.to_datetime("2023-07-01", utc=True).tz_convert("America/Sao_Paulo")) & (sp_data.index <= pd.to_datetime("2023-07-31", utc=True).tz_convert("America/Sao_Paulo"))]
sp_filtered_data_august =sp_data[(sp_data.index >= pd.to_datetime("2023-08-01", utc=True).tz_convert("America/Sao_Paulo")) & (sp_data.index <= pd.to_datetime("2023-08-31", utc=True).tz_convert("America/Sao_Paulo"))]


ibov_filtered_data_july = data[(data.index >= pd.to_datetime("2023-07-01", utc=True).tz_convert("America/Sao_Paulo")) & (data.index <= pd.to_datetime("2023-07-31", utc=True).tz_convert("America/Sao_Paulo"))]
ibov_filtered_data_august =data[(data.index >= pd.to_datetime("2023-08-01", utc=True).tz_convert("America/Sao_Paulo")) & (data.index <= pd.to_datetime("2023-08-31", utc=True).tz_convert("America/Sao_Paulo"))]

average_close_sp_july = sp_filtered_data_july['Close'].mean()
average_close_sp_august = sp_filtered_data_august['Close'].mean()

average_close_ibov_july = ibov_filtered_data_july['Close'].mean()
average_close_ibov_august = ibov_filtered_data_august['Close'].mean()

sp_variation_july_to_august = (average_close_sp_august - average_close_sp_july) / average_close_sp_july * 100
ibov_variation_july_to_august = (average_close_ibov_august - average_close_ibov_july) / average_close_ibov_july * 100

print("S&P 500 - July", average_close_sp_july)
print("S&P 500 - August", average_close_sp_august)
print("Variation July to August", sp_variation_july_to_august)

print("IBOVESPA - July", average_close_ibov_july)
print("IBOVESPA - August", average_close_ibov_august)
print("Variation July to August", ibov_variation_july_to_august)

S&P 500 - July 4503.818410773027
S&P 500 - August 4455.072243430398
Variation July to August -1.082329767692877
IBOVESPA - July 119121.2
IBOVESPA - August 117686.45454545454
Variation July to August -1.2044417404672325


AI SAID IBOVESPA WAS **129,266.5** REALITY SAID **117,686.45**

AI SAID IBOVESPA ROSE **4.7%** REALITY SAID IT FALL **1.2%**

AI SAID S&P ROSE **3,1%** RELATITY SAID IT FALL **1.08%**

ALL THE NUMBERS WERE WRONG, AS THE ANALYSIS.